### PR TITLE
New Action: Salesforce Campaigns

### PR DESCRIPTION
**Pull request by davidtamaki**
This Action allows Looker users to create campaigns in their Salesforce organization and populate the campaign members with leads and/or contacts. 

One new package has been added to manage the Salesforce API: [jsforce](https://github.com/jsforce/jsforce)

Following the same pattern as the google actions, this action also uses the AH oauth functionality and will thus require SALESFORCE_CLIENT_ID and SALESFORCE_CLIENT_SECRET to be passed as environment variables*. The action will not register itself with the hub if they are not present.

Future work will be to add streaming results functionality to support larger result sets.

Looking forward to any and all feedback!

*UPDATE - after PR feedback, this action will also require SALESFORCE_MAX_RESULTS and SALESFORCE_CHUNK_SIZE as env variables.  Max results is max number of objects to fetch. This is used in the form builder to get existing campaigns (default is 10,000). Chunk size is the number of sObject sent per single request (limit is currently 200 records)

### CODE REVIEW

File: src/actions/salesforce/campaigns/campaigns_form_builder.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+export class SalesforceCampaignsFormBuilder {
+  readonly oauthCreds: { oauthClientId: string; oauthClientSecret: string }
+
+  constructor(oauthClientId: string, oauthClientSecret: string) {
+    this.oauthCreds = { oauthClientId, oauthClientSecret }
+  }
+
+  async formBuilder(request: Hub.ActionRequest, tokens: Tokens) {
+    // const sfdcConn = await salesforceLogin(request);
```
Comment (theo-browne) [inline src/actions/salesforce/campaigns/campaigns_form_builder.ts:14]: remove commented out line

File: src/actions/salesforce/campaigns/campaigns_send_data.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    this.oauthCreds = { oauthClientId, oauthClientSecret }
+  }
+
+  async sendData(
+    request: Hub.ActionRequest,
+    memberIds: MemberIds[],
+    tokens: Tokens,
+  ) {
+    // const sfdcConn = await salesforceLogin(request);
+    const sfdcConn = await sfdcConnFromRequest(request, tokens, this.oauthCreds)
```
Comment (theo-browne) [inline src/actions/salesforce/campaigns/campaigns_send_data.ts:34]: It looks like this block is repeated from above. Likely worth moving into its own function.

File: src/actions/salesforce/campaigns/campaigns_send_data.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+          throw new Error(
+            `Campaign creation error: ${JSON.stringify(newCampaign.errors)}`,
+          )
+        } else {
+          campaignId = newCampaign.id
+        }
+
+        break
+      case "append":
+        campaignId = request.formParams.campaign_name!
```
Comment (theo-browne) [inline src/actions/salesforce/campaigns/campaigns_send_data.ts:63]: break

File: src/actions/salesforce/campaigns/campaigns_send_data.ts [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+      column.data.map((id) => {
+        return {
+          [column.sfdcMemberType]: id,
+          CampaignId: campaignId,
+          Status: request.formParams.member_status,
+        }
+      }),
+    )
+
+    // flattan array of arrays into one big list
```
Comment (theo-browne) [inline src/actions/salesforce/campaigns/campaigns_send_data.ts:77]: typo

File: src/actions/salesforce/campaigns/campaigns_send_data.ts [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+      const cleanErrors = memberErrors.map((me) => {
+        const memberMessage = me.errors.errors.map((e: any) => e.message)
+        return { [me.memberId]: memberMessage }
+      })
+      const summary = `Errors with ${memberErrors.length} out of ${memberCount} members`
+      message = `${summary}. ${JSON.stringify(cleanErrors)}`
+      winston.debug(message)
+    }
+
+    return { tokens: newTokens, message }
```
Comment (theo-browne) [inline src/actions/salesforce/campaigns/campaigns_send_data.ts:128]: Probably worth splitting most of these blocks into their own functions

File: src/actions/salesforce/campaigns/salesforce_campaigns.ts [hunk 6]
```diff
@@ -0,0 +1,255 @@
+import * as winston from "winston"
+import * as Hub from "../../../hub"
+import { SalesforceOauthHelper } from "../common/oauth_helper"
+import { SalesforceCampaignsFormBuilder } from "./campaigns_form_builder"
+import { SalesforceCampaignsSendData } from "./campaigns_send_data"
+
+export const REDIRECT_URL = `${process.env.ACTION_HUB_BASE_URL}/actions/salesforce_campaigns/oauth_redirect`
+export const MAX_RESULTS = 10000 // how many existing campaigns to retrieve to append members to
```
Comment (theo-browne) [inline src/actions/salesforce/campaigns/salesforce_campaigns.ts:8]: Can we make these (MAX_RESULTS, CHUNK_SIZE) ENV variables

File: src/actions/salesforce/campaigns/test_salesforce_campaigns.ts [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+
+    it("returns an error response if there is no tokens", async () => {
+      request.attachment = { dataJSON: validLookerData }
+      const expectedResponse = makeErrorResponse({message: "Request state is missing code and redirect"})
+      const actualResponse = await action.validateAndExecute(request)
+
+      expect(actualResponse).to.deep.equal(expectedResponse)
+    })
+
+    // it("returns an error response if domain is invalid", async () => {
```
Comment (theo-browne) [inline src/actions/salesforce/campaigns/test_salesforce_campaigns.ts:168]: delete if unused

File: src/actions/salesforce/campaigns/salesforce_campaigns.ts [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+  async execute(request: Hub.ActionRequest) {
+    if (!(request.attachment && request.attachment.dataJSON)) {
+      throw "No attached json."
+    }
+
+    if (!(request.formParams.campaign_name)) {
+      throw "Missing Salesforce campaign name."
+    }
+
+    const qr = request.attachment.dataJSON
```
Comment (theo-browne) [inline src/actions/salesforce/campaigns/salesforce_campaigns.ts:98]: Change variable name to something more descriptive

**Comment by phillipperalez** [comment]
Does Salesforce use infinite tokens or long term tokens for oauth? Do they expire?

**Comment by phillipperalez** [comment]
Also apologies, you got caught in the jump to Node 14

File: src/actions/salesforce/campaigns/test_salesforce_campaigns.ts [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+const validLookerData = {
+    fields: { dimensions: [{ label: "Salesforce Contact Id", tags: ["sfdc_contact_id"], type: "string", name: "salesforce.contact_id" }]},
+    data: [{ "salesforce.contact_id": { value: "abc" } }, { "salesforce.contact_id": { value: "def" } }],
+}
+
+const action = new SalesforceCampaignsAction(testOAuthClientId, testOAuthClientSecret)
+
+describe(`${action.constructor.name} class`, () => {
+  // Use a sandbox to avoid interaction with other test files
+  const adsSinonSandbox = sinon.createSandbox()
```
Comment (theo-browne) [inline src/actions/salesforce/campaigns/test_salesforce_campaigns.ts:24]: Typo?

**Comment by davidtamaki** [comment]
@phillipperalez The access token has a configurable validity period (default is 2 minutes), and there is an option to use a [refresh token](https://help.salesforce.com/s/articleView?id=remoteaccess_oauth_refresh_token_flow.htm&type=5&language=en_US) to request a new access token when the access token expires. The refresh tokens do not expire.

This action is set up with the refresh_token scope.

File: src/actions/salesforce/campaigns/campaigns_send_data.ts (hunk 2 above)
Comment (davidtamaki) [reply src/actions/salesforce/campaigns/campaigns_send_data.ts:31]: Turns out that listening for the on refresh event is unnecessary and I can just pass back the connection object which handles the refresh automatically. I've moved this logic to very end of both functions as the refresh event can occur at any time.

**Comment by davidtamaki** [comment]
@theo-browne I've updated the PR with the your feedback comments and also for the node 14 update. Would you be able to take a final look?

**Review by theo-browne — APPROVED** [review]
LGTM

**Pull request merged**